source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -316,6 +310,98 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
   current_request_ = client->client_->makeRequest(ClusterSlotsRequest::instance_, *this);
 }
 
+void RedisCluster::RedisDiscoverySession::updateDnsStats(
+    Network::DnsResolver::ResolutionStatus status, bool empty_response) {
+  if (status == Network::DnsResolver::ResolutionStatus::Failure || empty_response) {
+    if (status == Network::DnsResolver::ResolutionStatus::Failure) {
+      parent_.info_->stats().update_failure_.inc();
+    } else {
+      parent_.info_->stats().update_empty_.inc();
+    }
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::resolveClusterHostnames(ClusterSlotsPtr&& slots) {
+  std::shared_ptr<ClusterSlotsPtr> s = std::make_shared<ClusterSlotsPtr>(std::move(slots));
+  resolveHostname(s, 0);
+}
+
+void RedisCluster::RedisDiscoverySession::resolveHostname(std::shared_ptr<ClusterSlotsPtr> slots,
+                                                          std::size_t index) {
+  if (index >= (**slots).size()) {
+    finishClusterHostnameResolution(std::move(slots));
+    return;
+  }
+  auto& slot = (**slots)[index];
+  if (slot.primary() == nullptr) {
+    ENVOY_LOG(trace,
+              "starting async DNS resolution for primary slot address {} at index location {}",
+              slot.primary_hostname_, index);
+    parent_.dns_resolver_->resolve(
+        slot.primary_hostname_, parent_.dns_lookup_family_,
+        [this, index,
+         slots = std::move(slots)](Network::DnsResolver::ResolutionStatus status,
+                                   std::list<Network::DnsResponse>&& response) -> void {
+          auto& slot = (**slots)[index];
+          ENVOY_LOG(trace, "async DNS resolution complete for {}", slot.primary_hostname_);
+          updateDnsStats(status, response.empty());
+          // If DNS resolution for a primary fails, we stop resolution for remaining, and reset the
+          // timer.
+          if (status != Network::DnsResolver::ResolutionStatus::Success) {
+            ENVOY_LOG(error, "Unable to resolve cluster slot primary hostname {}",
+                      slot.primary_hostname_);
+            resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+            return;
+          }
+          // Primary slot address resolved
+          slot.setPrimary(Network::Utility::getAddressWithPort(
+              *response.front().addrInfo().address_, slot.primary_port_));
+          // Continue on to resolve replicas
+          resolveReplicas(std::move(slots), index);
+        });
+  } else {
+    resolveReplicas(std::move(slots), index);
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::resolveReplicas(std::shared_ptr<ClusterSlotsPtr> slots,
```

Comment:
Why pass a shared ptr to a unique ptr here? Can we just pass the ClusterSlots by ref?

Reply:
Let me preface by saying I'm not familiar enough with C++ to know the best approach. 😅 

Initially, the whole logic had the rvalue-reference of `ClusterSlotsPtr&&`, but that was causing segfaults as the memory was getting cleaned up before it was accessed inside the DNS resolution lambda.

So, to ensure the ownership of the `cluster_slots` existed inside the lambda, I needed to make a copy of it inside the capture list. In order to do that safely, I'm using a `shared_ptr` instead of the `unique_ptr`. (well, the compiler wouldn't allow me to use the unique ptr anyway)

Reply:
Can we change ClustersSlotsPtr into a shared pointer instead of wrapping it? I think that would let you share ownership without this extra indirection?

Reply:
Looking at the code it doesn't seem to be widely used, so I think it's fine to just update those sites. I think I'd prefer increasing the scope of this PR slightly over the extra indirection, which feels like a bit of a code smell

Reply:
Sounds good. I implemented the change.